### PR TITLE
Make it possible to delete or restart a dynamic Shovel on/from any cluster node

### PR DESCRIPTION
**Pull request by kostakal**
## Proposed Changes

Currently, shovels on nodes not the target node that the user is connected to are not able to be deleted. see issue: 
https://github.com/rabbitmq/rabbitmq-server/issues/2807

This change adds functionality to locate node that shovel is running on, and call rpc:call to delete/restart shovel on that node.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [ ] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [ ] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.

### CODE REVIEW

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl [hunk 1]
```diff
@@ -107,52 +99,28 @@ filter_vhost_req(List, ReqData) ->
...
-      ReqData, Context).
-
-status(Node) ->
-    case rpc:call(Node, rabbit_shovel_status, status, [], infinity) of
-        {badrpc, {'EXIT', _}} ->
-            [];
-        Status ->
-            [format(Node, I) || I <- Status]
+restart_shovel(VHost, Name, Node) ->
+    rabbit_log:info("Calling rabbit_shovel_util:restart_shovel on Node '~s' for shovel '~s' in vhost '~s'", [Node, Name, VHost]),
```
Comment (michaelklishin) [inline deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl:103]: This is probably a development logging leftover?

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl [hunk 2]
```diff
@@ -107,52 +99,28 @@ filter_vhost_req(List, ReqData) ->
...
+    case rpc:call(Node, rabbit_shovel_util, restart_shovel, [VHost, Name], infinity) of
```
Comment (michaelklishin) [inline deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl:104]: `infinity` timeouts is a very bad idea. We have years of experience to back this claim.

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl [hunk 3]
```diff
@@ -107,52 +99,28 @@ filter_vhost_req(List, ReqData) ->
...
+        ok -> true;
```
Comment (michaelklishin) [inline deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl:105]: It's not common to return a boolean in Erlang to indicate success of an operation. `ok | {error, term()}` is used pretty much everywhere.

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt_util.erl [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                         VHost     -> lists:member(VHost, VHosts)
+                     end].
+
+status(ReqData, Context) ->
+    filter_vhost_user(
+        lists:append([status(Node) || Node <- [node() | nodes()]]),
+        ReqData, Context).
+
+status(Node) ->
+    case rpc:call(Node, rabbit_shovel_status, status, [], infinity) of
```
Comment (michaelklishin) [inline deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt_util.erl:34]: Same thing as above, never use `infinity` timeouts, in particular when calling across the network.

**Review by michaelklishin — CHANGES_REQUESTED** [review]
This needs some more work.

I don't know if an integration test would be useful to have (it can be too involved to justify,
e.g. detecting what node the shovel will land on can be too intrusive)
but testing individual filtering functions with mocks doesn't cover much.

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl (hunk 3 above)
Comment (kostakal) [reply deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl:105]: the api reply is true/false, however i can move that up to the reply level

Comment (michaelklishin) [reply deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl:105]: Yes, that can be perfectly reasonable in an API response.

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl (hunk 1 above)
Comment (michaelklishin) [reply deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl:103]: "Calling function X" will not make any sense to most users. So at least it should say something else, e.g.
"Asked to restart shovel … in virtual host … on node …"

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.hrl [hunk 5]
```diff
@@ -0,0 +1,8 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2021 VMware, Inc. or its affiliates.  All rights reserved.
+%%
+
+-define(SHOVEL_CALLS_TIMEOUT_MS, 5000).
```
Comment (michaelklishin) [inline deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.hrl:8]: 5 seconds is an absolute minimum. Both in RabbitMQ itself and client libraries, 15s is the absolute minimum we would use, in particular when the value is a constant.

**Review by michaelklishin — CHANGES_REQUESTED** [review]

File: deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl (hunk 3 above)
Comment (michaelklishin) [reply deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl]: As far as I can tell this has not been addressed.

Comment (kostakal) [reply deps/rabbitmq_shovel_management/src/rabbit_shovel_mgmt.erl]: hmm i moved it to part of the Reply assignment now to return from API - do you mean the way the error is handled?

**Review by michaelklishin — COMMENTED** [review]
So far I could not get this to work for CLI tool commands and restarts over the HTTP API.
I'll try to dive deeper later in the day.

**Comment by kostakal** [comment]
> So far I could not get this to work for CLI tool commands and restarts over the HTTP API. I'll try to dive deeper later in the day.

hm, for http api delete is working and restart is not? i can try again on my side as well - last time i did all seemed good. I did not check with cli tool commands, what kind of cli tool commands are you using?

**Comment by michaelklishin** [comment]
FTR, I am still working on the CLI part of this and some refactoring to make them share some relevant code.

**Comment by michaelklishin** [comment]
Moved to https://github.com/rabbitmq/rabbitmq-server/pull/4242.

**Pull request closed without merging**